// @jet-app/app-store/tmp/src/api/models/reviews.js
/**
 * Created by phare on 3/28/17.
 */
import * as base from "./base";
import * as page from "./generic-page";
/* tslint:disable:no-unsafe-any */
/** @public */
export class ReviewsPageSortOption extends base.Model {
    constructor(id, title, selectedActionTitle, url) {
        super();
        this.id = id;
        this.title = title;
        this.selectedActionTitle = selectedActionTitle;
        this.url = url;
    }
}
/** @public */
export class ReviewsPage extends page.GenericPage {
    constructor() {
        super([]);
    }
}
/** @public */
export class Ratings extends base.Model {
}
/** @public */
export class Review extends base.ViewModel {
}
/** @public */
export class Response extends base.Model {
}
/** @public */
export class TapToRate extends base.Model {
    constructor() {
        super(...arguments);
        this.componentType = "tapToRate";
    }
}
/** @public */
export class ProductReviewActions extends base.Model {
    constructor() {
        super(...arguments);
        this.componentType = "productReviewActions";
    }
}
/** @public */
export class EditorsChoice extends base.Model {
    constructor(notes) {
        super();
        this.showsBadge = false;
        this.title = null;
        this.notes = notes;
        this.isCollapsed = true;
    }
}
/** @public */
export class ReviewsContainer extends base.Model {
}
/** @public */
export class ProductReviewAction extends base.Model {
}
/** @public */
export class ProductWriteAReview extends base.Model {
}
/** @public */
export class ProductStarRatings extends Ratings {
    constructor() {
        super(...arguments);
        this.componentType = "starRatings";
    }
}
/** @public */
export class ProductStarRatingsHistogram extends Ratings {
    constructor() {
        super(...arguments);
        this.componentType = "starRatingsHistogram";
    }
}
/** @public */
export class ProductNoRatings extends Ratings {
    constructor() {
        super(...arguments);
        this.componentType = "noRatings";
    }
}
/** @public */
export class ProductRatingsAndReviewsMessage extends base.Model {
    constructor(messageText) {
        super();
        this.componentType = "message";
        this.messageText = messageText;
    }
}
/** @public */
export class ReviewSummary extends base.ViewModel {
    constructor(body, bodyNoTitle, subtitle, subtitleArtwork, subtitleArtworkAlignment, bodyMediaType, flowPreviewActionsConfiguration) {
        super();
        this.body = body;
        this.bodyNoTitle = bodyNoTitle;
        this.subtitle = subtitle;
        this.subtitleArtwork = subtitleArtwork;
        this.subtitleArtworkAlignment = subtitleArtworkAlignment;
        this.bodyMediaType = bodyMediaType;
        this.flowPreviewActionsConfiguration = flowPreviewActionsConfiguration;
    }
}
/** @public */
export class ProductReview extends base.ViewModel {
    constructor() {
        super(...arguments);
        this.componentType = "productReview";
    }
}
/* tslint:enable:no-unsafe-any */
