// @jet-app/app-store/tmp/src/api/models/shelf-based-product.js
/**
 * Created by km on 2/13/17.
 */
import * as dynamicGenericPage from "./dynamic-generic-page";
export const shelfBasedProductPageOrderingIdPurcahsed = "purchasedOrdering";
export const shelfBasedProductPageOrderingIdNotPurcahsed = "notPurchasedOrdering";
export const shelfBasedProductPageOrderingIdPurcahsedExpandedMedia = "purchasedOrdering_ExpandedMedia";
export const shelfBasedProductPageOrderingIdNotPurcahsedExpandedMedia = "notPurchasedOrdering_ExpandedMedia";
export const shelfBasedProductPageOrderingIdPurcahsedCompact = "purchasedOrdering_Compact";
export const shelfBasedProductPageOrderingIdNotPurcahsedCompact = "notPurchasedOrdering_Compact";
export const shelfBasedProductPageOrderingIdPurcahsedExpandedMediaCompact = "purchasedOrdering_ExpandedMedia_Compact";
export const shelfBasedProductPageOrderingIdNotPurcahsedExpandedMediaCompact = "notPurchasedOrdering_ExpandedMedia_Compact";
export const shelfBasedProductPageOrderingIdDownloading = "downloadingOrdering";
export const shelfBasedProductPageOrderingIdDownloadingExpandedMedia = "downloadingOrdering_ExpandedMedia";
export const shelfBasedProductPageOrderingIdDownloadingCompact = "downloadingOrdering_Compact";
export const shelfBasedProductPageOrderingIdDownloadingExpandedMediaCompact = "downloadingOrdering_ExpandedMedia_Compact";
/** @public */
export class ShelfBasedProductPage extends dynamicGenericPage.DynamicGenericPage {
    constructor() {
        super(...arguments);
        // Reviews
        this.appPlatforms = [];
    }
}
