// @jet-app/app-store/tmp/src/common/search/metadata-ribbon/metadata-ribbon-item-factory.js
import * as categoryMetadataRibbonItem from "./category-metadata-ribbon-item";
import * as chartMetadataRibbonItem from "./chart-metadata-ribbon-item";
import * as developerMetadataRibbonItem from "./developer-metadata-ribbon-item";
import * as dividerMetadataRibbonItem from "./divider-metadata-ribbon-item";
import * as editorsChoiceMetadataRibbonItem from "./editors-choice-metadata-ribbon-item";
import * as gameControllerMetadataRibbonItem from "./game-controller-metadata-ribbon-item";
import * as secondaryShortCategoriesMetadataRibbonItem from "./secondary-short-categories-metadata-ribbon-item";
import * as shortCategoryMetadataRibbonItem from "./short-category-metadata-ribbon-item";
import * as starRatingMetadataRibbonItem from "./star-rating-metadata-ribbon-item";
import * as tagMetadataRibbonItem from "./tag-metadata-ribbon-item";
export const standardList = {
    // every key in Key must be present
    chartPositions: chartMetadataRibbonItem.createMetadataRibbonItems,
    genreDisplayName: categoryMetadataRibbonItem.createMetadataRibbonItems,
    genreShortDisplayName: shortCategoryMetadataRibbonItem.createMetadataRibbonItems,
    secondaryGenreShortDisplayNames: secondaryShortCategoriesMetadataRibbonItem.createMetadataRibbonItems,
    developerInfo: developerMetadataRibbonItem.createMetadataRibbonItems,
    editorialBadgeInfo: editorsChoiceMetadataRibbonItem.createMetadataRibbonItems,
    userRating: starRatingMetadataRibbonItem.createMetadataRibbonItems,
    supportsGameController: gameControllerMetadataRibbonItem.createMetadataRibbonItems,
    tag: tagMetadataRibbonItem.createMetadataRibbonItems,
    divider: dividerMetadataRibbonItem.createMetadataRibbonItems,
};
