// @jet-app/app-store/tmp/src/gameservicesui/src/foundation/media-api/requests/recommendation-request-types.js
export const AllPlayNowFeedContent = [
    "players",
    "apps",
    "achievements",
    "leaderboards",
    "contents",
    "challenges",
    "challenge-instances",
    "activities",
];
export const AllPlayNowFeedContentTypes = [
    "friend-request-modules",
    "friend-playing-modules",
    "achievement-modules",
    "leaderboard-modules",
    "chart-modules",
    "content-modules",
    "challenge-invite-modules",
    "challenge-modules",
    "challenge-suggestion-modules",
    "on-device-modules",
    "activities",
    "activity-invite-modules",
    "activity-modules",
    "activity-suggestion-modules",
];
export const PlayNowFeedAppsAttributes = [
    "compatibilityControllerRequirement",
    "editorialArtwork",
    "editorialClientParams",
    "editorialVideo",
    "expectedReleaseDateDisplayFormat",
    "isAppleWatchSupported",
    "minimumOSVersion",
    "screenshotsByType",
    "videoPreviewsByType",
    "gameDisplayName",
    "miniGamesDeepLink",
];
export const PlayNowFeedEditorialItemsAttributes = [
    "editorialArtwork",
    "editorialClientParams",
    "editorialVideo",
    "expectedReleaseDateDisplayFormat",
    "showExpectedReleaseDate",
];
export const PlayNowFeedEditorialPagesAttributes = ["editorialArtwork", "editorialClientParams", "editorialVideo"];
export const AppEventsAttributes = ["description", "productArtwork", "productVideo"];
