// @jet-app/app-store/tmp/src/api/typings/constants.js
import { makeMetatype } from "@jet/environment/util/metatype";
export const storeMetrics = makeMetatype("app-store:storeMetrics");
export const amsEngagement = makeMetatype("app-store:amsEngagement");
export const device = makeMetatype("app-store:device");
export const user = makeMetatype("app-store:user");
export const player = makeMetatype("games:player");
export const metricsIdentifiers = makeMetatype("app-store:metricsIdentifiers");
export const arcade = makeMetatype("app-store:arcade");
export const gameCenter = makeMetatype("app-store:gameCenter");
export const resilientDeepLinks = makeMetatype("app-store:resilientDeepLinks");
export const ads = makeMetatype("app-store:ads");
export const onDeviceRecommendationsManager = makeMetatype("app-store:onDeviceRecommendationsManager");
export const onDeviceSearchHistoryManager = makeMetatype("app-store:onDeviceSearchHistoryManager");
export const featureFlags = makeMetatype("app-store:featureFlags");
export const mediaToken = makeMetatype("app-store:mediaTokenService");
export const adsLocalizer = makeMetatype("app-store:adsLocalizer");
export const appDistribution = makeMetatype("app-store:appDistribution");
export const timeoutManager = makeMetatype("app-store:timeoutManager");
export const treatmentStore = makeMetatype("app-store:treatmentStore");
export const userDefaults = makeMetatype("app-store:userDefaults");
