// @jet-app/app-store/tmp/src/common/product-page/shelf-based/product-page-shelf-ordering/product-page-shelf-ordering-ios.js
import { ProductPageSectionMapping } from "../product-page-section-mapping";
/// The at which to place a section that is in the spotlight.
export const purchasedSpotlightIndex = 5;
export const notPurchasedSpotlightIndex = 5;
// MARK: - Installed
export const purchasedOrdering = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", [
        "productRatings",
        "allProductReviews",
        "allProductReviewActions",
        "productReviewsFooter",
    ]),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "purchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
    new ProductPageSectionMapping("shelf", "similarItems"),
];
export const purchasedOrderingCompact = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", [
        "productRatings",
        "reviewSummary",
        "userProductReviews",
        "editorsChoiceProductReviews",
        "productReviewsFooter",
        "tapToRateProductReviewAction",
        "writeAReviewProductReviewAction",
    ]),
    new ProductPageSectionMapping("shelf", "tapToRateProductReviewAction"),
    new ProductPageSectionMapping("shelf", "userProductReviews"),
    new ProductPageSectionMapping("shelf", "editorsChoiceProductReviews"),
    new ProductPageSectionMapping("shelf", "writeAReviewProductReviewAction"),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "purchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
    new ProductPageSectionMapping("shelf", "similarItems"),
];
// MARK: - Not Installed
/// The standard ordering to use for an app that is not installed.
export const notPurchasedOrdering = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", ["productRatings", "allProductReviews", "productReviewsFooter"]),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "notPurchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
    new ProductPageSectionMapping("shelf", "similarItems"),
];
export const notPurchasedOrderingCompact = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", [
        "productRatings",
        "reviewSummary",
        "userProductReviews",
        "editorsChoiceProductReviews",
        "productReviewsFooter",
    ]),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "notPurchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
    new ProductPageSectionMapping("shelf", "similarItems"),
];
// MARK: - During Download
export const downloadingOrdering = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "similarItems"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", [
        "productRatings",
        "allProductReviews",
        "allProductReviewActions",
        "productReviewsFooter",
    ]),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "purchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
];
export const downloadingOrderingCompact = [
    new ProductPageSectionMapping("header", "header"),
    new ProductPageSectionMapping("topLockup", "topLockup"),
    new ProductPageSectionMapping("shelf", "informationRibbon"),
    new ProductPageSectionMapping("shelf", "productPageInlineMessage"),
    new ProductPageSectionMapping("shelf", "preorderDisclaimer"),
    new ProductPageSectionMapping("shelf", "similarItems"),
    new ProductPageSectionMapping("shelf", "bundleChildren"),
    new ProductPageSectionMapping("shelf", "bundleParents"),
    new ProductPageSectionMapping("shelf", "appEvents"),
    new ProductPageSectionMapping("shelf", "mostRecentVersion"),
    new ProductPageSectionMapping("shelf", "subscriptions"),
    new ProductPageSectionMapping("shelf", "inAppPurchases"),
    new ProductPageSectionMapping("shelf", "achievements"),
    new ProductPageSectionMapping("shelf", "friendsPlaying"),
    new ProductPageSectionMapping("shelf", "videos"),
    new ProductPageSectionMapping("screenshots", "screenshots"),
    new ProductPageSectionMapping("shelf", "description"),
    new ProductPageSectionMapping("shelf", "editorialQuote"),
    new ProductPageSectionMapping("shelf", "editorsChoice"),
    new ProductPageSectionMapping("shelf", "reviews", [
        "productRatings",
        "reviewSummary",
        "userProductReviews",
        "editorsChoiceProductReviews",
        "productReviewsFooter",
        "tapToRateProductReviewAction",
        "writeAReviewProductReviewAction",
    ]),
    new ProductPageSectionMapping("shelf", "tapToRateProductReviewAction"),
    new ProductPageSectionMapping("shelf", "userProductReviews"),
    new ProductPageSectionMapping("shelf", "editorsChoiceProductReviews"),
    new ProductPageSectionMapping("shelf", "writeAReviewProductReviewAction"),
    new ProductPageSectionMapping("shelf", "privacyHeader"),
    new ProductPageSectionMapping("shelf", "privacyTypes"),
    new ProductPageSectionMapping("shelf", "privacyFooter"),
    new ProductPageSectionMapping("shelf", "accessibilityHeader"),
    new ProductPageSectionMapping("shelf", "accessibilityFeatures"),
    new ProductPageSectionMapping("shelf", "accessibilityDeveloperLink"),
    new ProductPageSectionMapping("shelf", "information"),
    new ProductPageSectionMapping("shelf", "purchasedLinks"),
    new ProductPageSectionMapping("shelf", "capabilities"),
    new ProductPageSectionMapping("shelf", "featuredIn"),
    new ProductPageSectionMapping("shelf", "moreByDeveloper"),
];
