// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/category-detail-motion-16x9.js
import { ArtworkContentMode, EdgeInsetsZero, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const categoryDetailMotion16x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.bottom,
    expandedContentMode: ArtworkContentMode.bottom,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedLayoutInsets: EdgeInsetsZero,
};
const categoryDetailMotion16x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.top,
    expandedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedLayoutInsets: EdgeInsetsZero,
};
const categoryDetailMotion16x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: categoryDetailMotion16x9Layout_mini,
    rtl: categoryDetailMotion16x9Layout_mini,
    collapsedSize: { type: "absolute", width: 455, height: 256 },
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const categoryDetailMotion16x9Metrics_base = {
    ltr: categoryDetailMotion16x9Layout_base,
    rtl: categoryDetailMotion16x9Layout_base,
    collapsedSize: { type: "fractionalHeight", height: 1.0 },
};
export const categoryDetailMotion16x9 = {
    objectPath: "editorialVideo.categoryDetailMotion16x9",
    cardArtLayoutMetrics: [categoryDetailMotion16x9Metrics_mini, categoryDetailMotion16x9Metrics_base],
    sourceWidth: 800,
    sourceHeight: 450,
    type: "video",
    crops: ["sr"],
};
