// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/list-card-static-21x9.js
import { ArtworkContentMode, EdgeInsetsZero, Size, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const listCardStatic21x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.bottomRight,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.topRight,
    expandedLayoutInsets: EdgeInsetsZero,
};
const listCardStatic21x9Layout_mini_rtl = {
    collapsedContentMode: ArtworkContentMode.bottomLeft,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.topLeft,
    expandedLayoutInsets: EdgeInsetsZero,
};
const listCardStatic21x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: -49,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.right,
    expandedLayoutInsets: EdgeInsetsZero,
};
const listCardStatic21x9Layout_base_rtl = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: -49,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.left,
    expandedLayoutInsets: EdgeInsetsZero,
};
const listCardStatic16x9Layout_extraWide = {
    collapsedContentMode: ArtworkContentMode.scaleAspectFill,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.bottomRight,
    expandedLayoutInsets: {
        top: 50,
        left: 495,
        bottom: 0,
        right: 0,
    },
};
const listCardStatic16x9Layout_extraWide_rtl = {
    collapsedContentMode: ArtworkContentMode.scaleAspectFill,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.bottomRight,
    expandedLayoutInsets: {
        top: 50,
        left: -495,
        bottom: 0,
        right: 0,
    },
};
const listCardStatic21x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: listCardStatic21x9Layout_mini,
    rtl: listCardStatic21x9Layout_mini_rtl,
    sourceCropOverrideLTR: "LCS.ApLCS01",
    sourceCropOverrideRTL: "LCS.ApLCS02",
    sourceSizeOverride: new Size(550, 264),
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const listCardStatic21x9Metrics_base = {
    ltr: listCardStatic21x9Layout_base,
    rtl: listCardStatic21x9Layout_base_rtl,
    collapsedSize: { type: "absolute", width: 672, height: 279 },
};
const listCardMotion16x9Metrics_extraWide = {
    ltr: listCardStatic16x9Layout_extraWide,
    rtl: listCardStatic16x9Layout_extraWide_rtl,
    expandedSize: { type: "absolute", width: 1124, height: 482 },
    sourceCropOverrideLTR: "LCS.ApLCXW01",
    sourceCropOverrideRTL: "LCS.ApLCXW01",
    priority: TodayCardArtworkSizedLayoutMetricsPriority.ExtraWide,
};
export const listCardStatic21x9 = {
    objectPath: "editorialArtwork.listCardStatic21x9",
    cardArtLayoutMetrics: [
        listCardStatic21x9Metrics_mini,
        listCardStatic21x9Metrics_base,
        listCardMotion16x9Metrics_extraWide,
    ],
    crops: ["LCS.ApLCL01", "LCS.ApLCL02"],
    sourceWidth: 688,
    sourceHeight: 286,
    type: "image",
};
