// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/list-card-motion-21x9.js
import { ArtworkContentMode, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const expandedContentMode = ArtworkContentMode.topRight;
const expandedLayoutInsets = {
    top: -108,
    left: 495,
    bottom: 0,
    right: 0,
};
const expandedContentMode_rtl = ArtworkContentMode.topLeft;
const expandedLayoutInsets_rtl = {
    top: -108,
    left: -495,
    bottom: 0,
    right: 0,
};
const expandedSize = { type: "absolute", width: 1124, height: 482 };
const listCardMotion16x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -117,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode,
    expandedLayoutInsets,
};
const listCardMotion16x9Layout_mini_rtl = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -117,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: expandedContentMode_rtl,
    expandedLayoutInsets: expandedLayoutInsets_rtl,
};
const listCardMotion16x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: -165,
        left: 524,
        bottom: 0,
        right: 0,
    },
    expandedContentMode,
    expandedLayoutInsets,
};
const listCardMotion16x9Layout_base_rtl = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: -165,
        left: -524,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: expandedContentMode_rtl,
    expandedLayoutInsets: expandedLayoutInsets_rtl,
};
const listCardMotion16x9Layout_extraWide = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -55,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode,
    expandedLayoutInsets,
};
const listCardMotion16x9Layout_extraWide_rtl = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -55,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: expandedContentMode_rtl,
    expandedLayoutInsets: expandedLayoutInsets_rtl,
};
const listCardMotion16x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: listCardMotion16x9Layout_mini,
    rtl: listCardMotion16x9Layout_mini_rtl,
    collapsedSize: { type: "absolute", width: 914, height: 392 },
    expandedSize,
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const listCardMotion16x9Metrics_base = {
    ltr: listCardMotion16x9Layout_base,
    rtl: listCardMotion16x9Layout_base_rtl,
    collapsedSize: { type: "absolute", width: 1188, height: 509 },
    expandedSize,
};
const listCardMotion16x9Metrics_extraWide = {
    ltr: listCardMotion16x9Layout_extraWide,
    rtl: listCardMotion16x9Layout_extraWide_rtl,
    collapsedSize: { type: "absolute", width: 1456, height: 624 },
    expandedSize,
    priority: TodayCardArtworkSizedLayoutMetricsPriority.ExtraWide,
};
export const listCardMotion21x9 = {
    objectPath: "editorialVideo.listCardMotion21x9",
    cardArtLayoutMetrics: [
        listCardMotion16x9Metrics_mini,
        listCardMotion16x9Metrics_base,
        listCardMotion16x9Metrics_extraWide,
    ],
    crops: [],
    sourceWidth: 1208,
    sourceHeight: 518,
    type: "video",
};
