// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/story-card-static-16x9.js
import { ArtworkContentMode, EdgeInsetsZero, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const storyCardStatic16x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.bottom,
    expandedContentMode: ArtworkContentMode.bottom,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedLayoutInsets: EdgeInsetsZero,
};
const storyCardStatic16x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.bottom,
    expandedContentMode: ArtworkContentMode.bottom,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedLayoutInsets: EdgeInsetsZero,
};
const storyCardStatic16x9Layout_700w = {
    collapsedContentMode: ArtworkContentMode.scaleAspectFill,
    expandedContentMode: ArtworkContentMode.bottom,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedLayoutInsets: EdgeInsetsZero,
};
const storyCardStatic16x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: storyCardStatic16x9Layout_mini,
    rtl: storyCardStatic16x9Layout_mini,
    collapsedSize: { type: "absolute", width: 455, height: 256 },
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const storyCardStatic16x9Metrics_base = {
    maxWidth: 704,
    ltr: storyCardStatic16x9Layout_base,
    rtl: storyCardStatic16x9Layout_base,
    collapsedSize: { type: "fractionalHeight", height: 1.08 },
};
const storyCardStatic16x9Metrics_700w = {
    ltr: storyCardStatic16x9Layout_700w,
    rtl: storyCardStatic16x9Layout_700w,
};
export const storyCardStatic16x9 = {
    objectPath: "editorialArtwork.storyCardStatic16x9",
    cardArtLayoutMetrics: [
        storyCardStatic16x9Metrics_mini,
        storyCardStatic16x9Metrics_base,
        storyCardStatic16x9Metrics_700w,
    ],
    crops: ["sr"],
    sourceWidth: 800,
    sourceHeight: 450,
    type: "image",
};
