// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/universal-a-static-16x9.js
import { ArtworkContentMode, EdgeInsetsZero, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const universalAStatic16x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: 16,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.scaleAspectFill,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.scaleAspectFill,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_700w = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: 0,
        left: -70,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_700w_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: 0,
        left: 70,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_920w = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_920w_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_max = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Layout_max_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: EdgeInsetsZero,
    expandedContentMode: ArtworkContentMode.center,
    expandedLayoutInsets: EdgeInsetsZero,
};
const universalAStatic16x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: universalAStatic16x9Layout_mini,
    rtl: universalAStatic16x9Layout_mini,
    collapsedSize: { type: "absolute", width: 455, height: 256 },
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const universalAStatic16x9Metrics_base = {
    maxWidth: 699.0,
    ltr: universalAStatic16x9Layout_base,
    rtl: universalAStatic16x9Layout_base,
};
const universalAStatic16x9Metrics_700w = {
    maxWidth: 899.0,
    ltr: universalAStatic16x9Layout_700w,
    rtl: universalAStatic16x9Layout_700w_rtl,
};
const universalAStatic16x9Metrics_920w = {
    maxWidth: 1090.0,
    ltr: universalAStatic16x9Layout_920w,
    rtl: universalAStatic16x9Layout_920w_rtl,
    collapsedSize: { type: "absolute", width: 1400, height: 527 },
};
const universalAStatic16x9Metrics_max = {
    ltr: universalAStatic16x9Layout_max,
    rtl: universalAStatic16x9Layout_max_rtl,
    collapsedSize: { type: "absolute", width: 1600, height: 604 },
};
export const universalAStatic16x9 = {
    objectPath: "editorialArtwork.universalAStatic16x9",
    cardArtLayoutMetrics: [
        universalAStatic16x9Metrics_mini,
        universalAStatic16x9Metrics_base,
        universalAStatic16x9Metrics_700w,
        universalAStatic16x9Metrics_920w,
        universalAStatic16x9Metrics_max,
    ],
    crops: ["UAS.ApXWC01"],
    sourceWidth: 1600,
    sourceHeight: 604,
    type: "image",
};
