// @jet-app/app-store/tmp/src/common/today/artwork/media-configurations/universal-a-motion-16x9.js
import { ArtworkContentMode, TodayCardArtworkSizedLayoutMetricsPriority, } from "../../../../api/models";
const expandedSize = { type: "absolute", width: 1200, height: 675 };
const expandedLayoutInsets = {
    top: -130,
    left: 0,
    bottom: 0,
    right: 0,
};
const universalAMotion16x9Layout_mini = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -16,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_base = {
    collapsedContentMode: ArtworkContentMode.top,
    collapsedLayoutInsets: {
        top: -116,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_700w = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: -116,
        left: -70,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_700w_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: -116,
        left: 70,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_920w = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: -148,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_920w_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: -148,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_max = {
    collapsedContentMode: ArtworkContentMode.topLeft,
    collapsedLayoutInsets: {
        top: -170,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Layout_max_rtl = {
    collapsedContentMode: ArtworkContentMode.topRight,
    collapsedLayoutInsets: {
        top: -170,
        left: 0,
        bottom: 0,
        right: 0,
    },
    expandedContentMode: ArtworkContentMode.top,
    expandedLayoutInsets,
};
const universalAMotion16x9Metrics_mini = {
    maxWidth: 250.0,
    ltr: universalAMotion16x9Layout_mini,
    rtl: universalAMotion16x9Layout_mini,
    collapsedSize: { type: "absolute", width: 455, height: 256 },
    expandedSize,
    priority: TodayCardArtworkSizedLayoutMetricsPriority.Mini,
};
const universalAMotion16x9Metrics_base = {
    maxWidth: 699.0,
    ltr: universalAMotion16x9Layout_base,
    rtl: universalAMotion16x9Layout_base,
    collapsedSize: { type: "absolute", width: 1094, height: 614 },
    expandedSize,
};
const universalAMotion16x9Metrics_700w = {
    maxWidth: 899.0,
    ltr: universalAMotion16x9Layout_700w,
    rtl: universalAMotion16x9Layout_700w_rtl,
    collapsedSize: { type: "absolute", width: 1092, height: 614 },
    expandedSize,
};
const universalAMotion16x9Metrics_920w = {
    maxWidth: 1090.0,
    ltr: universalAMotion16x9Layout_920w,
    rtl: universalAMotion16x9Layout_920w_rtl,
    collapsedSize: { type: "absolute", width: 1392, height: 783 },
    expandedSize,
};
const universalAMotion16x9Metrics_max = {
    ltr: universalAMotion16x9Layout_max,
    rtl: universalAMotion16x9Layout_max_rtl,
    collapsedSize: { type: "absolute", width: 1600, height: 900 },
    expandedSize,
};
export const universalAMotion16x9 = {
    objectPath: "editorialVideo.universalAMotion16x9",
    cardArtLayoutMetrics: [
        universalAMotion16x9Metrics_mini,
        universalAMotion16x9Metrics_base,
        universalAMotion16x9Metrics_700w,
        universalAMotion16x9Metrics_920w,
        universalAMotion16x9Metrics_max,
    ],
    crops: ["UAS.ApXWC01"],
    sourceWidth: 1600,
    sourceHeight: 900,
    type: "video",
};
